plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/ui/LoginServiceUi.java

```diff
@@ -67,22 +74,60 @@ public void run() {
 
   @Override
   public VerificationCodeHolder obtainVerificationCodeFromExternalUserInteraction(String title) {
-    throw new RuntimeException("Not to be called."); //$NON-NLS-1$
+    AuthorizationCodeReceiver codeReceiver = new AuthorizationCodeReceiver();
+
+    String redirectUrl;
+    try {
+      redirectUrl = codeReceiver.runLocalServer();
+    } catch (IOException ex) {
+      showErrorDialogHelper(Messages.LOGIN_ERROR_DIALOG_TITLE,
+          Messages.LOGIN_ERROR_LOCAL_SERVER_INIT
+              + ex.getLocalizedMessage());
+      return null;
+    }
+
+    try {
+      if (!Program.launch(GoogleLoginService.getGoogleLoginUrl(redirectUrl))) {
+        showErrorDialogHelper(
+            Messages.LOGIN_ERROR_DIALOG_TITLE, Messages.LOGIN_ERROR_CANNOT_OPEN_BROWSER);
+        return null;
+      }
+
+      showProgressDialogAndWaitForCode(codeReceiver);
+      return new VerificationCodeHolder(codeReceiver.getAuthorizationCode(), redirectUrl);
+
+    } catch (InvocationTargetException ex) {
+      showErrorDialogHelper(Messages.LOGIN_ERROR_DIALOG_TITLE,
+          Messages.LOGIN_ERROR_LOCAL_SERVER_RUN
+              + ex.getLocalizedMessage());
+      return null;
+    } catch (InterruptedException ex) {
+      return null;  // User pushed the cancel button.
+    }
+    finally {
+      codeReceiver.shutdown();
+    }
+  }
+
+  private void showProgressDialogAndWaitForCode(final AuthorizationCodeReceiver codeReceiver)
+      throws InvocationTargetException, InterruptedException {
+    new ProgressMonitorDialog(shellProvider.getShell()).run(true /* fork */, true /* cancelable */,
+        new IRunnableWithProgress() {
+          @Override
+          public void run(IProgressMonitor monitor)
+              throws InvocationTargetException, InterruptedException {
+            while (!codeReceiver.waitForCode(500 /* timeout: 0.5 second */)) {
+              if (monitor.isCanceled()) {
+                throw new InterruptedException();
+              }
+            }
+          }
+        });
   }
 
   @Override
   public String obtainVerificationCodeFromUserInteraction(
       String title, GoogleAuthorizationCodeRequestUrl authCodeRequestUrl) {
-    if (!Program.launch(authCodeRequestUrl.toString())) {
-      MessageDialog.openError(shellProvider.getShell(),
-          Messages.LOGIN_ERROR_DIALOG_TITLE, Messages.LOGIN_ERROR_CANNOT_OPEN_BROWSER);
-      return null;
-    }
-
-    // TODO(chanseok): remove InputDialog and set up a web server to receive authorization code.
-    InputDialog dialog = new InputDialog(null,
-        "Enter Authorization Code", "Enter authorization code from the browser.", null, null);
-    dialog.open();
-    return dialog.getValue();
+    throw new RuntimeException("Not to be called."); //$NON-NLS-1$
```

Comment:
Not needed for both IntelliJ and Eclipse, since this method is only for entering an authorization code manually. The IDEs won't use it, but I think it makes sense for the method to exist in the ide-login library that covers all cases.